### PR TITLE
Use TypeIdMap whenever possible

### PR DESCRIPTION
**Pull request by SpecificProtagonist**
Use `TypeIdMap<T>` instead of `HashMap<TypeId, T>`

- ~~`TypeIdMap` was in `bevy_ecs`. I've kept it there because of #11478~~
- ~~I haven't swapped `bevy_reflect` over because it doesn't depend on `bevy_ecs`, but I'd also be happy with moving `TypeIdMap` to `bevy_utils` and then adding a dependency to that~~
- ~~this is a slight change in the public API of `DrawFunctionsInternal`, does this need to go in the changelog?~~

## Changelog
- moved `TypeIdMap` to `bevy_utils`
- changed `DrawFunctionsInternal::indices` to `TypeIdMap`

## Migration Guide

- `TypeIdMap` now lives in `bevy_utils`
- `DrawFunctionsInternal::indices` now uses a `TypeIdMap`.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Yeah, we should note the change.

I think we should move this into bevy_utils for now: it definitely doesn't belong in bevy_ecs and moving it out might help us see other more natural groupings rather than a catch-all utility crate.

**Review by NthTensor — APPROVED** [review]
Very happy to see this moved into `bevy_util`. Nice work.

**Comment by alice-i-cecile** [comment]
Added a migration guide and resolved merge conflicts. Attempting merge, fingers crossed I formatted the import statement correctly 😅

File: crates/bevy_gizmos/src/lib.rs [hunk 1]
```diff
@@ -77,7 +77,7 @@ use bevy_render::{
     renderer::RenderDevice,
     Extract, ExtractSchedule, Render, RenderApp, RenderSet,
 };
-use bevy_utils::HashMap;
+use bevy_utils::{tracing::warn, HashMap, TypeIdMap};
```
Comment (alice-i-cecile) [inline crates/bevy_gizmos/src/lib.rs:80]:
```suggestion
use bevy_utils::TypeIdMap;
```

**Comment by alice-i-cecile** [comment]
No worries, this is why we have CI :)

**Pull request merged**